### PR TITLE
Add test-report workflow

### PR DESCRIPTION
**Pull request by milesgranger**
- [x] Will close https://github.com/dask/dask/issues/10732
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

Checks out the distributed repo, and runs the `test_report.py` script to generate the report for dask/dask.

Depends on https://github.com/dask/distributed/pull/8450

### CODE REVIEW

**Comment by milesgranger** [comment]
Tested on my fork: 
https://github.com/milesgranger/dask/pull/3

Successful deployment workflow:
https://github.com/milesgranger/dask/actions/runs/7524967710/job/20480569721

And resulting page:
https://milesgranger.github.io/dask/test_report.html

File: .github/workflows/test-report.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            test_report*
+            !test_report.html
+          key: ${{ env.PYTHON_VERSION }}-${{ hashFiles('continuous_integration/scripts/test_report*') }}
+
+      - name: Generate report
+        run: |
+          python continuous_integration/scripts/test_report.py --repo dask/dask --max-days 90 --max-runs 30 --nfails 1 -o test_report.html
+          python continuous_integration/scripts/test_report.py --repo dask/dask --max-days 7 --max-runs 30 --nfails 2 -o test_short_report.html --title "Test Short Report"
+          mkdir deploy
+          mv test_report.html test_short_report.html deploy/
```
Comment (milesgranger) [inline .github/workflows/test-report.yaml:61]: For this to work, we need https://github.com/dask/distributed/pull/8461 in. 
The successful run on my fork modified `--max-days` to only recent days which didn't have the bad naming of artifacts.

File: .github/workflows/test-report.yaml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+jobs:
+  test-report:
+    name: Test Report
+    # Do not run the report job on forks
+    if: github.repository == 'dask/dask' || github.event_name == 'workflow_dispatch'
+    runs-on: ubuntu-latest
+    env:
+      GITHUB_TOKEN: ${{ github.token }}
+      PYTHON_VERSION: 3.9
```
Comment (crusaderky) [inline .github/workflows/test-report.yaml:20]:
```suggestion
```

File: .github/workflows/test-report.yaml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+      - uses: actions/cache@v3
+        id: cache
+        with:
+          # Suffix is depending on the backend / OS. Let's be agnostic here
+          # See https://docs.python.org/3/library/shelve.html#shelve.open
+          path: |
+            test_report*
+            !test_report.html
+          key: ${{ env.PYTHON_VERSION }}-${{ hashFiles('continuous_integration/scripts/test_report*') }}
```
Comment (crusaderky) [inline .github/workflows/test-report.yaml:54]:
```suggestion
                    # Note: these files are in the dask/distributed repository
                    key: ${{ hashFiles('continuous_integration/scripts/test_report*') }}
```

File: .github/workflows/test-report.yaml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+          repository: dask/distributed
+
+      - name: Setup Conda Environment
+        uses: conda-incubator/setup-miniconda@v3.0.1
+        with:
+          miniforge-variant: Mambaforge
+          miniforge-version: latest
+          condarc-file: continuous_integration/condarc
+          use-mamba: true
+          python-version: ${{ env.PYTHON_VERSION }}
```
Comment (crusaderky) [inline .github/workflows/test-report.yaml:36]:
```suggestion
          python-version: 3.9
```

File: .github/workflows/test-report.yaml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+          environment-file: continuous_integration/scripts/test-report-environment.yml
```
Comment (crusaderky) [inline .github/workflows/test-report.yaml:37]:
```suggestion
          # Note: this file is in the dask/distributed repo
          environment-file: continuous_integration/scripts/test-report-environment.yml
```

File: .github/workflows/test-report.yaml (hunk 3 above)
Comment (crusaderky) [reply .github/workflows/test-report.yaml:54]: Do I understand correctly that the cache is specific to dask/dask, so we don't need to match the key in dask/distributed?

**Comment by crusaderky** [comment]
XREF https://github.com/dask/distributed/pull/8462

File: .github/workflows/test-report.yaml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+          repository: dask/distributed
+
+      - name: Setup Conda Environment
+        uses: conda-incubator/setup-miniconda@v3.0.1
+        with:
+          miniforge-variant: Mambaforge
+          miniforge-version: latest
+          condarc-file: continuous_integration/condarc
+          use-mamba: true
+          python-version: 3.9
```
Comment (crusaderky) [inline .github/workflows/test-report.yaml:35]:
```suggestion
```
https://github.com/dask/distributed/pull/8462

File: .github/workflows/test-report.yaml (hunk 3 above)
Comment (milesgranger) [reply .github/workflows/test-report.yaml]: That's my understanding as well: 
https://github.com/milesgranger/dask/actions/runs/7524967710/job/20480569721#step:14:2

**Review by crusaderky — APPROVED** [review]

**Pull request merged**